### PR TITLE
imu_spoofer

### PR DESCRIPTION
**Pull request by marija-p**
- Changed imu_compass_asl to account for NED -> ENU transform
- Changed GPS conversion nodes to publish point message
- Changed GPS spoofer to publish point message
- Added subscriber to laser altitude on GPS spoofer (tested on Colibri)
- Removed code relevant to pressure sensor (not used anymore)
- Resolved conflicts with master

@enricgalceran

### CODE REVIEW

File: src/gps_to_pose_conversion_node.cpp [hunk 1]
```diff
@@ -76,6 +77,13 @@ void gps_callback(const sensor_msgs::NavSatFixConstPtr& msg)
...
   pose_msg->pose.pose.orientation = g_latest_imu_msg.orientation;
 
+  // Fill up position message
+  geometry_msgs::PointStampedPtr position_msg(
+    new geometry_msgs::PointStamped);
+  position_msg->header = pose_msg->header;
+  position_msg->header.frame_id = "world";
+  position_msg->point = pose_msg->pose.pose.position;
+
   if (g_got_pressure) {
```
Comment (enricgalceran) [inline src/gps_to_pose_conversion_node.cpp]: Please add the option of taking z from laser_altimeter, with a boolean parameter allowing to switch between gps altitude and laser altitude. Just add a subscriber that allows to take a Float64 as "external_altitude" input, keeping it general (as opposed to mentioning the laser altimeter).

File: src/gps_spoofer.py [hunk 2]
```diff
@@ -43,10 +53,12 @@ def __init__(self):
 
     self.pub_spoofed_gps = rospy.Publisher('spoofed_gps', NavSatFix, queue_size=1)
     self.pub_disturbed_pose = rospy.Publisher('disturbed_pose', PoseWithCovarianceStamped, queue_size=1,  tcp_nodelay=True)
+    self.pub_point = rospy.Publisher('vicon_point', PointStamped, queue_size=1,  tcp_nodelay=True)    
 
+    self.sub_altitude = rospy.Subscriber('laser_altitude', Float64, self.callback_altitude, tcp_nodelay=True)
     self.sub_gps = rospy.Subscriber('fcu/gps', NavSatFix, self.callback_gps, tcp_nodelay=True)
```
Comment (enricgalceran) [inline src/gps_spoofer.py]: Set topic name to simply 'gps' and remap in launch accordingly.

File: src/gps_spoofer.py [hunk 3]
```diff
@@ -43,10 +53,12 @@ def __init__(self):
...
     self.sub_estimated_odometry = rospy.Subscriber('estimated_odometry', Odometry, self.callback_odometry, queue_size=1, tcp_nodelay=True)
-    self.sub_pressure_height = rospy.Subscriber('pressure_height_point', PointStamped, self.callback_pressure_height, queue_size=1, tcp_nodelay=True)
+    self.sub_imu = rospy.Subscriber('fcu/imu', Imu, self.callback_imu, queue_size=1, tcp_nodelay=True)
```
Comment (enricgalceran) [inline src/gps_spoofer.py]: Same as for 'gps', rename to 'imu' and remap.

File: src/gps_spoofer.py [hunk 4]
```diff
@@ -58,21 +70,41 @@ def callback_odometry(self, data):
...
-    
-    # TODO: Take height measurement from pressure sensor
+
+    # Pose message
+#    self.pwc.pose.pose = data.pose.pose
+    x = data.pose.pose.position.x
+    y = data.pose.pose.position.y
+
+    #angle = radians(-7.0) #imu compass
+    angle = -2.73
```
Comment (enricgalceran) [inline src/gps_spoofer.py:83]: Add comment explaining what this angle thing does.

File: src/gps_spoofer.py [hunk 5]
```diff
@@ -58,21 +70,41 @@ def callback_odometry(self, data):
...
+
+    # Rotate x and y to align x to East
+    self.pwc.pose.pose.position.x = cos(angle)*x - sin(angle)*y
+    self.pwc.pose.pose.position.y = sin(angle)*x + cos(angle)*y
+
+    # TODO: Take height measurement from laser
```
Comment (enricgalceran) [inline src/gps_spoofer.py]: Is this still a TODO?

File: src/gps_spoofer.py [hunk 6]
```diff
@@ -58,21 +70,41 @@ def callback_odometry(self, data):
...
     self.pwc.pose.pose.position.z = data.pose.pose.position.z
+    self.pwc.pose.pose.orientation = self.latest_imu_message.orientation
+
+    # Point message
+    # Take x and y from vicon, take altitude from laser
+    self.point.header = self.pwc.header
+    self.point.point.x = x
+    self.point.point.y = y
+    self.point.point.z = self.altitude.data
```
Comment (enricgalceran) [inline src/gps_spoofer.py]: Add a config param allowing to switch altitude input between vicon Z and laser altitude.

File: src/gps_spoofer.py [hunk 7]
```diff
@@ -86,6 +118,10 @@ def publish_odometry(self, event):
     self.pwc.pose.pose.position.y += self.R_noise*sin(self.theta_noise)
```
Comment (enricgalceran) [inline src/gps_spoofer.py:129]: Not part of this PR, but please rename this method from 'publish_odometry' to something like 'publish_noisy_positions'.

File: src/gps_spoofer.py [hunk 8]
```diff
@@ -86,6 +118,10 @@ def publish_odometry(self, event):
...
     self.pub_disturbed_pose.publish(self.pwc)
 
+    self.point.point.x += self.R_noise*cos(self.theta_noise)
+    self.point.point.y +=self.R_noise*sin(self.theta_noise)
```
Comment (enricgalceran) [inline src/gps_spoofer.py]: Space after =.

File: src/imu_compass_asl.cpp [hunk 9]
```diff
@@ -311,6 +311,9 @@ void IMUCompass::repackageImuPublish(tf::StampedTransform transform)
 
   tf::Quaternion hackq = tf::createQuaternionFromRPY(hroll, hpitch, -(hyaw+M_PI/2.0));
 
+  // Change coordinate frame from NED to ENU for yaw velocity
```
Comment (enricgalceran) [inline src/imu_compass_asl.cpp]: Please just delete.

File: src/set_gps_reference_node.cpp [hunk 10]
```diff
@@ -38,7 +40,8 @@ void gps_callback(const sensor_msgs::NavSatFixConstPtr & msg)
       g_lat_ref /= g_its;
       g_lon_ref /= g_its;
       g_alt_ref /= g_its;
-    } else {
```
Comment (enricgalceran) [inline src/set_gps_reference_node.cpp:41]: Is that the standard formatting kicking in? (Really not sure, please check.)

File: src/gps_to_pose_conversion_node.cpp (hunk 1 above)
Comment (marija-p) [inline src/gps_to_pose_conversion_node.cpp]: I've added a subscriber for "external_altitude" input, replacing the one we had originally for pressure only. In this code, the altitude is published in the GPS position message, provided the node receives messages on the topic. Basically, if a remapping to "external_altitude" exists, the node will use this altitude; if not, it will take the GPS altitude. I think a Boolean parameter would be a bit redundant in this case, as we can simply switch between the two by controlling the topic remapping. What do you think, what would be the more elegant way of doing this?

File: src/set_gps_reference_node.cpp (hunk 10 above)
Comment (marija-p) [inline src/set_gps_reference_node.cpp:41]: Seems like this is standard. https://google.github.io/styleguide/cppguide.html

File: src/gps_spoofer.py (hunk 6 above)
Comment (marija-p) [inline src/gps_spoofer.py]: Okay, as this is a single param I'll add it to the launch file directly. Let me know if you think we need a .yaml.

**Comment by marija-p** [comment]
Cleaned up. @enricgalceran
